runtime/src/main/java/com/dylibso/chicory/runtime/Instance.java

```diff
@@ -129,73 +132,82 @@ public Instance initialize(boolean start) {
+                                    + " extra instructions");
+                }
                 Value offset = computeConstantValue(this, ae.offset());
                 if (offset.type() != ValueType.I32) {
-                    throw new ChicoryException("Invalid offset type in element " + offset.type());
+                    throw new InvalidException(
+                            "type mismatch, invalid offset type in element " + offset.type());
                 }
                 List<List<Instruction>> initializers = ae.initializers();
                 for (int i = 0; i < initializers.size(); i++) {
                     final List<Instruction> init = initializers.get(i);
                     var index = offset.asInt() + i;
+                    if (init.stream().filter(e -> e.opcode() != OpCode.END).count() > 1l) {
+                        throw new InvalidException(
+                                "constant expression required, type mismatch, expected [] but found"
+                                        + " extra instructions");
+                    }
                     var value = computeConstantValue(this, init);
                     var inst = computeConstantInstance(this, init);
+                    if (value.type() != ae.type() || table.elementType() != ae.type()) {
+                        throw new InvalidException(
+                                "type mismatch, element type: "
+                                        + ae.type()
+                                        + ", table type: "
+                                        + table.elementType()
+                                        + ", value type: "
+                                        + value.type());
+                    }
                     if (ae.type() == ValueType.FuncRef) {
-                        function(value.asFuncRef());
+                        try {
+                            function(value.asFuncRef());
+                        } catch (InvalidException e) {
+                            throw new InvalidException("type mismatch, " + e.getMessage(), e);
+                        }
                         table.setRef(index, value.asFuncRef(), inst);
                     } else {
                         assert ae.type() == ValueType.ExternRef;
                         table.setRef(index, value.asExtRef(), inst);
                     }
                 }
+            } else if (el instanceof DeclarativeElement) {
+                var de = (DeclarativeElement) el;
+
+                List<List<Instruction>> initializers = de.initializers();
+                for (int i = 0; i < initializers.size(); i++) {
+                    final List<Instruction> init = initializers.get(i);
+                    var value = computeConstantValue(this, init);
+                    var func = function(value.asFuncRef());
+                    if (de.type() == ValueType.FuncRef && func != null) {
```

Comment:
Related to the above comment, should `function()` throw on invalid values rather than returning null? With the current code, an invalid function index here is ignored.